### PR TITLE
feat: add enforcement for invariants to commit checker

### LINKED ISSUE #1880: Check Invariants are respecting table features for write paths
# Description

Check Invariants are respecting table features for write paths, @MrPowers if you can assign this to me so others know.

**Use Case**

This is for proper 3,7 support from #1568 

**Related Issue(s)**

#1568

### PR DESCRIPTION
**Pull request by hntd187**
# Description
Added invariant checking in the commit checker.

# Related Issue(s)
#1568 closes #1880

# Documentation

### CODE REVIEW

File: crates/deltalake-core/src/operations/transaction/protocol.rs [hunk 1]
```diff
@@ -142,6 +142,25 @@ impl ProtocolChecker {
             })?;
         }
 
+        // Does the table have invariants?
+        let invariants = snapshot
+            .current_metadata()
+            .and_then(|meta| meta.schema.get_invariants().ok())
+            .unwrap_or_default()
+            .is_empty();
```
Comment (wjones127) [inline crates/deltalake-core/src/operations/transaction/protocol.rs:150]: `invariants = false` meaning "it does have invariants" does feel very intuitive. Could we reverse it, and rename so it's a little clearer?

```suggestion
        let has_invariants = !snapshot
            .current_metadata()
            .and_then(|meta| meta.schema.get_invariants().ok())
            .unwrap_or_default()
            .is_empty();
```

File: crates/deltalake-core/src/operations/transaction/protocol.rs [hunk 2]
```diff
@@ -142,6 +142,25 @@ impl ProtocolChecker {
...
+
+        // The table is at least version 7, has actual invariants and has the proper writer feature
+        // enabled
+        if snapshot.min_writer_version() >= 7
+            && !invariants
```
Comment (wjones127) [inline crates/deltalake-core/src/operations/transaction/protocol.rs:155]:
```suggestion
            && has_invariants
```

**Review by wjones127 — CHANGES_REQUESTED** [review]

File: crates/deltalake-core/src/operations/transaction/protocol.rs (hunk 1 above)
Comment (hntd187) [reply crates/deltalake-core/src/operations/transaction/protocol.rs]: done

File: crates/deltalake-core/src/operations/transaction/protocol.rs (hunk 2 above)
Comment (hntd187) [reply crates/deltalake-core/src/operations/transaction/protocol.rs]: also done

File: crates/deltalake-core/src/kernel/actions/types.rs [hunk 3]
```diff
@@ -274,6 +274,26 @@ pub enum WriterFeatures {
...
+            | WriterFeatures::DeletionVectors
+            | WriterFeatures::RowTracking
+            | WriterFeatures::TimestampWithoutTimezone
+            | WriterFeatures::DomainMetadata
+            | WriterFeatures::V2Checkpoint
+            | WriterFeatures::IcebergCompatV1 => 7,
+        }
+    }
+}
+
```
Comment (roeap) [inline crates/deltalake-core/src/kernel/actions/types.rs:296]: Are we using this somewhere? As per our discussion on Slack, not entirely sure how and where we would use this.

**Review by roeap — CHANGES_REQUESTED** [review]
There may have been a bit of miscommunication :).

The thing we need to decide is, if we do the invariant check or not - i.e. are invariants enabled. This is the case for `2 <= writer_version < 7` and if `writer_version == 7 & 'invariants' in writer_features`. And only if invariants are enabled we should inspect the schema to see if there are any invariants to validate.

One way to do this would be to make sure we only pass invariants to the [`DeltaDataChecker`](https://github.com/delta-io/delta-rs/blob/dd6b45362a14c0f127b32c4b81afc15d17f710d5/crates/deltalake-core/src/delta_datafusion/mod.rs#L1021-L1024) if they are enabled.

It also means we likely need to error if invariants are enabled and we have some defined, but the datafusion features and with that the checker is not available.

File: crates/deltalake-core/src/kernel/actions/types.rs (hunk 3 above)
Comment (hntd187) [reply crates/deltalake-core/src/kernel/actions/types.rs:296]: No, I think this was me speculating about what we might need for determining say minimum versions across the entire hashset of features. I don't have any reason to say it should stay so I'm fine removing it if you would like.

**Comment by hntd187** [comment]
According to the protocol invariants for 2<=>6 are always enabled, so you'd only care to check if you were 7 or higher. Are you able to disable them prior to writer feature checks?

**Comment by roeap** [comment]
> Are you able to disable them prior to writer feature checks?

No, as you said, between 2 and 6 they are always enabled. The main point is that it is not an error if they are configured on the schema, so we should not error, rather if if they are not enabled, we just kind of pretend they don't exist and do nothing w.r.t. to them ...

**Comment by hntd187** [comment]
I added this in check constraints anyways so this PR isn't needed anymore.

**Pull request closed without merging**